### ISSUE TITLE
Websocket connection was abnormally closed, logout! code: 1006

### ISSUE THREAD

**Issue by achechen** [Lothiraldan/saltpad#156]
I can log in but Saltpad does not show any data (cannot pull anything from salt-api) and after a while it logs out with "Websocket connection was abnormally closed, logout! Reason: , code: 1006" error message.

See the screenshot that contains the errors (chrome console)

![capture](https://cloud.githubusercontent.com/assets/10340542/14556264/9a42f8fc-02f7-11e6-9234-5313e284f095.PNG)

settings.json:

```
{
    "API_URL": "myserver.local:8000",
    "SECURE_HTTP": false,
    "templates": {
        "basic": {
            "description": "Basic template",
            "matcher": "glob",
            "target": "*",
            "moduleFunction": "test.fib",
            "arguments": {
                "num": 10
            }
        },

        "version": {
            "description": "Basic version",
            "matcher": "glob",
            "target": "Celeste",
            "moduleFunction": "test.version",
            "arguments": {}
        }
    },
    "EAUTH": "pam",
    "FLAVOUR": "rest_cherrypy",
    "PATH_PREFIX": "/"
}
```

/etc/salt/master:

```
external_auth:
  pam:
    username:
        - .*
        - '@runner'
        - '@wheel'

...

rest_cherrypy:
  port: 8000
  host: 0.0.0.0
  disable_ssl: true
  debug: tru
```e

**Comment by achechen** [comment]
Yeah, I've missed the information that I need to use single-app deployment because version 2015.8.8 is not supported yet. Sorry.

**Comment by achechen** [comment]
OK, I cannot figure out how to use single-app deployment. Can anyone point me to the right direction?

Saltpad is located at /opt/saltpad

Running httpd with the following virtualhost config:

```
<VirtualHost *:80>
  ServerName myserver.local
  ServerAdmin webmaster@example.com
  LogLevel warn
  DocumentRoot /opt/saltpad
  <Directory "/opt/saltpad">
    RewriteEngine On
    RewriteBase /
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_URI} !index
    RewriteRule (.*) /index.html [L]
    #FallbackResource /index.html
  </Directory>
  ErrorLog "/var/log/httpd/saltpad-error.log"
  CustomLog "/var/log/httpd/saltpad-access.log" combined
</VirtualHost>
```

I cannot figure out what I should enter into these path fields below:

```
rest_cherrypy:
  port: 8000
  static: /code/static
  static_path: /static
  app: /code/index.html
  app_path: /saltpad
```

and what to put as PATH_PREFIX in saltpad config.

**Comment by Lothiraldan** [comment]
Hi @achechen,

currently saltpad supports two deployment options.

 - The recommended one is to use a web server (like nginx or apache) to server saltpad files, this solution requires that salt-api supports CORS, which is not the case for any netapi (rest_cherrypy or rest_tornado) in the current stable version.
 - "Rest_cherrypy single-app page deployment", rest_cherrypy in the current stable version also allows to "embed" a single-app-page and bypass the CORS need. It's not recommended because it's harder to deploy but it's the only viable solution for current stable version if you use rest_cherrypy.

In your case, I think you try to use the stable version, so you correctly deployed saltpad via rest_cherrypy, but you need apache to proxy request to salt-api directly and not try to serve the saltpad files itself.

I hope it will helps you make saltpad works correctly.

**Comment by achechen** [comment]
OK, let me check if I understand you correctly.

I am running version 2015.8.8 and neither rest_tornado nor rest_cherrypy is not supported for this version and I need to use embedded rest_cherrypy, right?

In case of embedded rest_cherrypy, I do not need a webserver as cherrypy will be serving the files.

I've stopped httpd and using only cherrypy now.

salt master config is as follows:

```
rest_cherrypy:
  port: 8000
  host: 0.0.0.0
  static: /opt/saltpad/static
  static_path: /static
  app: /opt/saltpad/index.html
  app_path: /saltpad/
  disable_ssl: true
```

and saltpad config:

```
{
    "API_URL": "myserver.local:8000",
    "SECURE_HTTP": false,
    "templates": {},
    "EAUTH": "pam",
    "FLAVOUR": "rest_cherrypy",
    "PATH_PREFIX": "/saltpad/"
}
```

When I try to open http://myserver.local:8000 in my browser I get the following error displayed:

```
404 Not Found

The path '/saltpad/' was not found.

Powered by CherryPy 3.2.2
```

Apparently there is something wrong with the path. What am I doing wrong?

**Comment by achechen** [comment]
Ok, I've deployed the v0.3.1 version and now I am having the same issue as JaminZhang (Issue #158). So frustrating :(

**Comment by achechen** [comment]
Ok, I was finally able to run it. I'll explain how.

THIS IS SINGLE APP DEPLOYMENT!!!

First of all, don't use https://github.com/Lothiraldan/saltpad/releases/download/v0.3/dist.zip as it says in the README, use https://github.com/Lothiraldan/saltpad/releases/download/v0.3.1/dist.zip

Unzip it in /opt/, it will create a dist folder. Rename it as saltpad.

In /opt/saltpad/static, create settings.json (it is not there) with the following:

```
{
    "API_URL": "your.servers.external.adress:8000",
    "SECURE_HTTP": false,
    "templates": {},
    "EAUTH": "pam",
    "FLAVOUR": "rest_cherrypy",
    "PATH_PREFIX": "/saltpad"
}

```

Include the following lines in /etc/salt/master:

```
external_auth:
  pam:
    <your_username>:
        - .*
        - '@runner'
        - '@wheel'

rest_cherrypy:
  port: 8000
  static: /opt/saltpad/static
  static_path: /static
  app: /opt/saltpad/index.html
  app_path: /saltpad
  disable_ssl: true

```

Restart salt-api and salt-master.

Open a browser and go to the following link:

your.servers.external.adress:8000/saltpad/    - (**last slash is crucial**, otherwise you get "Warning: Location "/saltpad" did not match any routes")

log in and voila! :)

Please note that. minions list takes a while to load.

**Comment by JaminZhang** [comment]
Good job!!!
It works for me.

@Lothiraldan 
Is it safe to expose the salt-api port 8000 to the WAN network?

**Comment by JaminZhang** [comment]
@achechen @Lothiraldan 

When I relogin the saltpad, the minions list is empty.
And I can't run job either...
What about you? @achechen

**Comment by achechen** [comment]
Yeah, firs of all, you need to create a job template in settings.json. I have this:

```
{
    "API_URL": "your.server.address:8000",
    "SECURE_HTTP": false,
    "templates": {
        "basic": {
            "description": "Basic template",
            "matcher": "glob",
            "target": "*",
            "moduleFunction": "test.ping"
        }
    },
    "EAUTH": "pam",
    "FLAVOUR": "rest_cherrypy",
    "PATH_PREFIX": "/saltpad"
}
```

After that, reload the page and you will see the template in "Jobs templates" section where you can launch it. After that, the minion list will be populated.

**Comment by Lothiraldan** [comment]
Thank you very much for the debug process, I've already updated the README and will try to clarify the README.

About the "Run Job" view, you'll need to first launch a "salt -v '*' sys.list_functions test" if you want to have all the test functions listed in the Function dropdown. I plan to add the list of commonly available functions directly in saltpad but it may cause some troubles like with different salt minions versions or  Operating Systems on which the minions run.

Saltpad list minions by detecting when they responds to a Job, it will helps to launch a "salt -v '*' test.ping" to detect them all.

**Comment by JaminZhang** [comment]
@Lothiraldan 
Can saltpad keep the minion in the minions list after login out? 
Because when I logout and login again, the minion is no longer in the minions list.
That's not convenient.

**Comment by Lothiraldan** [comment]
@JaminZhang, the problem with saving the minion or the jobs in your browser when you logout is that it's much more easy to exploit a vulnerability in your browser to get these information than attacking the salt-api. I don't want to put any sensible information in your browser (either via cookie, session-storage or something else) because of the general security of browsers.

The only piece of informations saved when you either reload the browser or logout is the list of your 5 last launched jobs and favored jobs, and even in this case it's only the job id that is saved.

I'm working on consolidating the list of minions from all the possible sources of informations.

I'm putting a security section in the documentation about that.

**Comment by btbroot** [comment]
Hi. Can't get it working.

````
$ salt-master --version
salt-master 2016.3.0 (Boron)
$ salt-api --version
salt-api 2016.3.0 (Boron)
````
This version, in my understanding, should support CORS. At least I can see "CORS" mentioned in `rest_cherrypy/app.py`.

````
$ cat /etc/opt/saltpad/salt-master.conf
rest_cherrypy:
  port: 8000
  disable_ssl: true
external_auth:
  pam:
    admin%:
      - .*
      - '@runner'
      - '@wheel'
````
````
$ cat /etc/opt/saltpad/apache2-site.conf 
<VirtualHost _default_:444>
	DocumentRoot /opt/saltpad
	SSLEngine on
	SSLCertificateFile	/etc/ssl/certs/ssl-cert-snakeoil.pem
	SSLCertificateKeyFile /etc/ssl/private/ssl-cert-snakeoil.key
	<Directory /opt/saltpad>
		FallbackResource /index.html
		Require all granted
	</Directory>
	<Location /api/>
		ProxyPass http://localhost:8000/
	</Location>
</VirtualHost>
````
````
$ cat /etc/opt/saltpad/settings.json 
{
    "API_URL": "salt.xxxxx.xx:444",
    "SECURE_HTTP": true,
    "FLAVOUR": "rest_cherrypy"
}
````
Everything works as expected till pushing the "Login" button.

In Firefox I'm getting an error notification "e.data.return is undefined".

In Chromium: "Cannot read property '0' of undefined".

![2016-07-05 03 38 42](https://cloud.githubusercontent.com/assets/4640976/16571311/d5642e3a-426a-11e6-8ac9-a495656c44fc.png)

**Comment by btbroot** [comment]
I'll just leave it here.

As of yet the only working solution for SaltPad and stable SaltStack is the rest_cherrypy single application serving option, but recent cherrypy3 has built-in SSL support totally broken since a while. For anyone using consistent OS distributions that could be an obstacle.

For Debian GNU/Linux a temporary workaround would be to downgrade python-cherrypy3 package to the version shipped with oldstable (wheezy) and to pin it (possibly enabling the oldstable security repository). The old package can be rebuilt in stable environment just fine.
